pythonforandroid/recipe.py

```diff
@@ -725,14 +725,44 @@ class PythonRecipe(Recipe):
     This is almost always what you want to do.'''
 
     setup_extra_args = []
-    '''List of extra arugments to pass to setup.py'''
+    '''List of extra arguments to pass to setup.py'''
+
+    depends = [('python2', 'python3')]
+    '''
+    .. note:: it's important to keep this depends as a class attribute, outside
+              `__init__` because, sometimes, we only initialize the object, so
+              the `__init__` call it won't be called, which will lead to not
+              have the python versions as a dependencies and it will cause a
+              tremendous `test_graph` error (difficult to track) and also, the
+              build order for dependencies will not be computed as expected (if
+              computed...). So be very careful with this line!!
+
+    .. warning:: this `depends` may be overwrote in inherited classes of
+                 `PythonRecipe`, so we make sure that any sub class will
+                 contain python as a dependency. We do this by checking the
+                 dependencies in meth:`PythonRecipe.__init__` method and adding
+                 them again in case that is necessary, so don't forget to call
+                 `super` in any inherited class of this class.
+    '''
 
     def __init__(self, *args, **kwargs):
         super(PythonRecipe, self).__init__(*args, **kwargs)
-        depends = self.depends
-        depends.append(('python2', 'python3', 'python3crystax'))
-        depends = list(set(depends))
-        self.depends = depends
+        if not any(
+            [
+                d
+                for d in {'python2', 'python3', ('python2', 'python3')}
+                if d in self.depends
+            ]
+        ):
+            # we overwrote `depends` in inherited recipe, so we must add it
+            # again the python versions as dependencies, but we only do this in
+            # case that the sub classes recipe does not contain any python
+            # version as dependency because it may be some recipes only
+            # compatible with a single version of python
+            depends = self.depends
+            depends.append(('python2', 'python3'))
+            depends = list(set(depends))
+            self.depends = depends
```

Comment:
This looks like the wrong place to do it to me. I think this should be removed and handled in `pythonforandroid/graph.py` instead, because this would also ensure that not calling `super().__init__()` is never a problem. I it is "obvious" that any recipe is meant to use python so semantically I don't think it is important or relevant to enforce this here, and it's completely fine to fix this in graph.py since it maintains the spirit of the class if it's not in here (if that explanation makes any sense to you)